### ISSUE TITLE
Single Node.js instance

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#158]
As discussed at https://github.com/runtimejs/runtime/issues/75#issuecomment-120875129, implement a module to use a single instance of Node.js. Probably there's something similar available, but the idea here is to "replace" the Node.js executable itself so all the new executions are redirected to the same Node.js instance. Probably this will not be really useful until threated WebWorkers support don't arribe to io.js/Node.js-next, but until that some proof of concept can be done with child_process.fork creating new processes.

**Comment by piranna** [comment]
Another option is to use the [multi-context](https://strongloop.com/strongblog/whats-new-node-js-v0-12-multiple-context-execution/) feature, available from the [vm](https://nodejs.org/api/vm.html) module.

**Comment by formula1** [comment]
I'll see about making a gist tonight. It shouldn't be absurdly difficult to get a proof of concept

**Comment by piranna** [comment]
:+1: I did a quick search and only found pidfile module, but it's too primitive and focused on instance of an application and not node.js itself, so it's useless to us.

**Comment by formula1** [comment]
https://github.com/avz/node-mkfifo This is the sort of thing your were interested in correct?

Apparently using mkfifo is [dangerous territory](http://stackoverflow.com/questions/11750041/how-to-create-a-named-pipe-in-node-js). I'm currently implementing a simple version that just creates a version listening to a file.

As for the multi-context environment, I had [gotten started](https://github.com/NodeOS/NodeOS/issues/121) on that earlier. [Here](https://gist.github.com/formula1/c1aa2d81d4bb6b1b2dce#file-cancelablesandbox-js) is the most relevent file to it

According to strong loop, it seems as though net and other native modules are already being handled. However, when "closing" a context seems to still be a needed feature. As a result, keeping track of all asyncronous aspects created is pretty important and involves a lot more work : /. Threads appear to be a golden opportunity from a clean up perspective.

**Comment by piranna** [comment]
> https://github.com/avz/node-mkfifo This is the sort of thing your were interested in correct?
> 
> Apparently using mkfifo is dangerous territory. I'm currently implementing a simple version that just creates a version listening to a file.

Yes, that's what I was talking about :-) Didn't know about the flaws of mkfifo, but as in your second link is suggested, you can mostly do the same with domain sockets that are already supported by Node.js. I'm not sure about the diferences between them (maybe named pipes are lighter because they don't involve the network stack? maybe better compatibility of domain sockets with network-oriented applications?) but for this use case both are valid ones, so it's ok for me to use domain sockets :-)


> As for the multi-context environment, I had gotten started on that earlier. Here is the most relevent file to it

I read it a while ago but didn't understood too much about what you were propossing, sorry :-P I'll need to take it another look... :-)


> According to strong loop, it seems as though net and other native modules are already being handled. However, when "closing" a context seems to still be a needed feature. As a result, keeping track of all asyncronous aspects created is pretty important and involves a lot more work : /

The article is somewhat old (pre v0.12) so probably things got better now.


> Threads appear to be a golden opportunity from a clean up perspective.

Definitelly, I'm too much into processes but seems here is the right way to go :-)

**Comment by lite20** [comment]
Do we still want to attempt this?
